File: backend/app/api/routes/push.py
from fastapi import APIRouter, Depends, HTTPException, status, Request
from sqlalchemy.orm import Session
from uuid import UUID

from app.db.session import get_db
from app.api.deps import get_current_user, require_facility_access
from app.models.auth import User
from app.models.org import FacilityUserAccess
from app.schemas.push import (
    PushSubscribeRequest,
    PushUnsubscribeRequest,
    PushPublicKeyResponse,
    PushGenericResponse,
    PushPreferencesResponse,
    PushPreferencesUpdateRequest,
    PushTestRequest,
    PushTestResponse,
)
from app.services.push_service import (
    get_vapid_public_key,
    upsert_subscription,
    delete_subscription,
    get_push_preferences,
    upsert_push_preferences,
    send_test_push,
)

router = APIRouter(prefix="/push", tags=["push"])


def _require_admin_or_medico(db: Session, *, user: User, facility_id: UUID) -> None:
    if user.is_platform_admin:
        return
    access = (
        db.query(FacilityUserAccess)
        .filter(
            FacilityUserAccess.facility_id == facility_id,
            FacilityUserAccess.user_id == user.id,
            FacilityUserAccess.is_active.is_(True),
        )
        .first()
    )
    if not access:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="No tiene acceso a esta sede")
    if access.role not in ("ADMIN", "MEDICO"):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Solo ADMIN y MEDICO")


@router.get("/vapid-public-key", response_model=PushPublicKeyResponse)
async def vapid_public_key():
    try:
        return PushPublicKeyResponse(public_key=get_vapid_public_key())
    except RuntimeError:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail="Push no configurado")


@router.post("/subscribe", response_model=PushGenericResponse)
async def subscribe(
    payload: PushSubscribeRequest,
    request: Request,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    # Validar acceso al hogar
    require_facility_access(payload.facility_id)(current_user, db)

    ua = payload.user_agent or request.headers.get("user-agent")

    if not payload.endpoint or not payload.keys.p256dh or not payload.keys.auth:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Suscripción inválida")

    upsert_subscription(
        db,
        user_id=current_user.id,
        facility_id=payload.facility_id,
        endpoint=payload.endpoint,
        p256dh=payload.keys.p256dh,
        auth=payload.keys.auth,
        user_agent=ua,
    )

    return PushGenericResponse(message="Suscripción guardada")


@router.post("/unsubscribe", response_model=PushGenericResponse)
async def unsubscribe(
    payload: PushUnsubscribeRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    require_facility_access(payload.facility_id)(current_user, db)

    deleted = delete_subscription(
        db,
        user_id=current_user.id,
        facility_id=payload.facility_id,
        endpoint=payload.endpoint,
    )

    if deleted:
        return PushGenericResponse(message="Suscripción eliminada")
    return PushGenericResponse(message="No había suscripción")


@router.get("/preferences", response_model=PushPreferencesResponse)
async def get_preferences(
    facility_id: UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    require_facility_access(facility_id)(current_user, db)
    _require_admin_or_medico(db, user=current_user, facility_id=facility_id)

    disabled = get_push_preferences(db, user_id=current_user.id, facility_id=facility_id)
    return PushPreferencesResponse(facility_id=facility_id, disabled_event_types=disabled)


@router.post("/preferences", response_model=PushPreferencesResponse)
async def update_preferences(
    payload: PushPreferencesUpdateRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    require_facility_access(payload.facility_id)(current_user, db)
    _require_admin_or_medico(db, user=current_user, facility_id=payload.facility_id)

    pref = upsert_push_preferences(
        db,
        user_id=current_user.id,
        facility_id=payload.facility_id,
        disabled_event_types=payload.disabled_event_types,
    )
    disabled = pref.disabled_event_types if isinstance(pref.disabled_event_types, list) else []
    return PushPreferencesResponse(facility_id=payload.facility_id, disabled_event_types=disabled)


@router.post("/test", response_model=PushTestResponse)
async def test_push(
    payload: PushTestRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    require_facility_access(payload.facility_id)(current_user, db)
    _require_admin_or_medico(db, user=current_user, facility_id=payload.facility_id)

    result = send_test_push(db, user_id=current_user.id, facility_id=payload.facility_id)
    return PushTestResponse(
        attempted=int(result.get("attempted", 0)),
        delivered=int(result.get("delivered", 0)),
        deleted=int(result.get("deleted", 0)),
    )

File: frontend/src/api/push.ts
import { apiClient } from './client';

export interface PushPublicKeyResponse {
  public_key: string;
}

export interface PushSubscribeRequest {
  facility_id: string;
  endpoint: string;
  keys: {
    p256dh: string;
    auth: string;
  };
  user_agent?: string;
}

export interface PushUnsubscribeRequest {
  facility_id: string;
  endpoint: string;
}

export interface PushPreferencesResponse {
  facility_id: string;
  disabled_event_types: string[];
}

export interface PushPreferencesUpdateRequest {
  facility_id: string;
  disabled_event_types: string[];
}

export interface PushTestRequest {
  facility_id: string;
}

export interface PushTestResponse {
  attempted: number;
  delivered: number;
  deleted: number;
}

export const pushApi = {
  getVapidPublicKey: async (): Promise<string> => {
    const res = await apiClient.get<PushPublicKeyResponse>('/push/vapid-public-key');
    return res.public_key;
  },

  subscribe: async (payload: PushSubscribeRequest): Promise<{ message: string }> => {
    return apiClient.post<{ message: string }>('/push/subscribe', payload);
  },

  unsubscribe: async (payload: PushUnsubscribeRequest): Promise<{ message: string }> => {
    return apiClient.post<{ message: string }>('/push/unsubscribe', payload);
  },

  getPreferences: async (facilityId: string): Promise<PushPreferencesResponse> => {
    const sp = new URLSearchParams({ facility_id: facilityId });
    return apiClient.get<PushPreferencesResponse>(`/push/preferences?${sp.toString()}`);
  },

  updatePreferences: async (payload: PushPreferencesUpdateRequest): Promise<PushPreferencesResponse> => {
    return apiClient.post<PushPreferencesResponse>('/push/preferences', payload);
  },

  testPush: async (payload: PushTestRequest): Promise<PushTestResponse> => {
    return apiClient.post<PushTestResponse>('/push/test', payload);
  },
};

export function urlBase64ToUint8Array(base64String: string): Uint8Array<ArrayBuffer> {
  const padding = '='.repeat((4 - (base64String.length % 4)) % 4);
  const base64 = (base64String + padding).replace(/-/g, '+').replace(/_/g, '/');
  const rawData = window.atob(base64);
  const outputArray = new Uint8Array(rawData.length) as Uint8Array<ArrayBuffer>;
  for (let i = 0; i < rawData.length; ++i) {
    outputArray[i] = rawData.charCodeAt(i);
  }
  return outputArray;
}

export function urlBase64ToArrayBuffer(base64String: string): ArrayBuffer {
  const u8 = urlBase64ToUint8Array(base64String);
  // Asegurar que el ArrayBuffer sea exactamente del tamaño de la key
  const buf: ArrayBuffer = u8.buffer;
  return buf.slice(u8.byteOffset, u8.byteOffset + u8.byteLength);
}
